man/man1/mtools.1
.TH MTOOLS 1
.SH NAME
mtools \- tools to access FAT file systems
.SH SYNOPSIS
.B mtools
.RB [ \-V ]
.B msdos_command 
.RI [ \-msdos_options ]
.RI arguments " ..."
.SH DESCRIPTION
.de SP
.if t .sp 0.4
.if n .sp
..
.B Mtools
is a collection of utilities to access MS-DOS (FAT) disks from Unix without
mounting them. It supports the long filenames of Windows NT and Windows 95.
It does not support NTFS disks.
.P
Some versions of mtools for other operating systems provide separate
commands, such as mdir, mcopy, etc., to emulate similar MS-DOS and Windows
command line commands. The version ported to Minix takes the MS-DOS
command (dir, copy, etc.) as its first argument. Supported MS-DOS 
commands are:
.B attrib, 
.B badblocks, 
.B cat, 
.B cd, 
.B copy, 
.B del, 
.B deltree, 
.B dir,
.B doctorfat, 
.B du, 
.B format, 
.B info, 
.B label, 
.B md, 
.B mkdir, 
.B partition,
.B rd, 
.B rmdir, 
.B read, 
.B move, 
.B ren, 
.B showfat, 
.B type, 
.B write
.P
The MS-DOS options are the same as for DOS commands, except they are prefaced
with "-" instead of "\\".
.P
Use 'mtools msdos_command -?' for help per command. (This tells you "-?"
is an illegal command, but, as with Unix systems, entering an illegal command
often is the easiest way to find out what are the legal commands.)
.P
Note that a disk argument must be terminate by or separated from a path by 
a colon (":").
If no disk argument is given mtools assumes you meant "/dev/fd0:", the 
first floppy disk drive.
.SH OPTIONS
.TP
.B \-V
Show the mtools version and configuration
.SH EXAMPLES
.de EX
.TP 20
\\fB\\$1\\fR
# \\$2
..
.EX "mtools dir" "show directory of MS-DOS floppy in drive A:."
.EX "mtools copy /dev/c0d0p0:file.txt ." "copy file.txt from MS-DOS root directory to current Minix directory."
.SH "SEE ALSO"
.BR dosdir (1).
.BR dosread (1).
.BR doswrite (1).
.SH NOTES
.P
Mtools requires a lot of memory.  The default on a Minix 3 installation
is over 10 MB.  A default configured mtools would not run on a system
with only 16 MB RAM.  You may be able to make do by using chmem to
reduce the memory allocation of mtools. On the 16 MB system mentioned
mtools still works with a reduction of the memory allocation to half
the original value.  The amount of memory you need depends upon the
size of the MS-DOS or Windows file systems you want to access.
Typically systems with big disks also have large amounts of memory.  If
mtools won't work for you, you may be able to fall back to the old
dosdir, dosread, and doswrite Minix utilities if the FAT file system
you want to access is small enough (the dos* utilities can access FAT16
partitions up to 256 MB size).
.P
This man page does not attempt to be complete. A lot of information is
available on line. To use mtools well you also need to be familiar with
the options for the corresponding MS-DOS commands. 
For more information see the mtools website, http://mtools.linux.lu/.
.P
Mtools-3.9.10 was released on 1 March 2005.  The Minix port is of the
earlier Mtools version 3.9.7, dated 1 June 2000.
.P
The Minix port is configured with the following options: disable-xdf
disable-vold disable-new-vold disable-debug disable-raw-term (read the source
to understand what these mean).
.SH BUGS
Yes, bugs may exist, but as this man page is written we don't know of any.
Please report any you find.
.P
As with any program that accesses a foreign file system, reading is probably
safe, but you may want to experiment carefully before using these programs to
write to a Windows system. 
.SH AUTHOR
Mtools is maintained by David Niemi and Alain Knaff. 
.P
Ported to Minix 2.0.3 by Kees J. Bot <kjb@cs.vu.nl>.
.P
This man page compiled by Al Woodhull <asw@woodhull.com>. 
.\" rev 2006-06-17


commands/i386/mtools-3.9.7/mtools.c
#include "sysincludes.h"
#include "msdos.h"
#include "mtools.h"
#include "partition.h"
#include "vfat.h"

const char *progname;

static const struct dispatch {
	const char *cmd;
	void (*fn)(int, char **, int);
	int type;
} dispatch[] = {
	{"attrib",mattrib, 0},
	{"badblocks",mbadblocks, 0},
	{"cat",mcat, 0},
	{"cd",mcd, 0},
	{"copy",mcopy, 0},
	{"del",mdel, 0},
	{"deltree",mdel, 2},
	{"dir",mdir, 0},
	{"doctorfat",mdoctorfat, 0},
	{"du",mdu, 0},
	{"format",mformat, 0},
	{"info", minfo, 0},
	{"label",mlabel, 0},
	{"md",mmd, 0},
	{"mkdir",mmd, 0},
#ifdef OS_linux
	{"mount",mmount, 0},
#endif
	{"partition",mpartition, 0},
	{"rd",mdel, 1},
	{"rmdir",mdel, 1},
	{"read",mcopy, 0},
	{"move",mmove, 0},
	{"ren",mmove, 1},
	{"showfat", mshowfat, 0},
#ifndef NO_CONFIG
	{"toolstest", mtoolstest, 0},
#endif
	{"type",mcopy, 1},
	{"write",mcopy, 0},
#ifndef OS_Minix
	{"zip", mzip, 0}
#endif
};
#define NDISPATCH (sizeof dispatch / sizeof dispatch[0])

int main(int argc,char **argv)
{
	const char *name;
	int i;

	init_privs();
#ifdef __EMX__
	_wildcard(&argc,&argv);
#endif

/*#define PRIV_TEST*/

#ifdef PRIV_TEST
	{ 
		int euid;
		char command[100];
	
		printf("INIT: %d %d\n", getuid(), geteuid());
		drop_privs();
		printf("DROP: %d %d\n", getuid(), geteuid());
		reclaim_privs();
		printf("RECLAIM: %d %d\n", getuid(), geteuid());
		euid = geteuid();
		if(argc & 1) {
			drop_privs();
			printf("DROP: %d %d\n", getuid(), geteuid());
		}
		if(!((argc-1) & 2)) {
			destroy_privs();
			printf("DESTROY: %d %d\n", getuid(), geteuid());
		}
		sprintf(command, "a.out %d", euid);
		system(command);
		return 1;
	}
#endif


#ifdef __EMX__
       _wildcard(&argc,&argv);
#endif 


	/* check whether the compiler lays out structures in a sane way */
	if(sizeof(struct partition) != 16 ||
	   sizeof(struct directory) != 32 ||
	   sizeof(struct vfat_subentry) !=32) {
		fprintf(stderr,"Mtools has not been correctly compiled\n");
		fprintf(stderr,"Recompile it using a more recent compiler\n");
		return 137;
	}

#ifdef __EMX__
       argv[0] = _getname(argv[0]); _remext(argv[0]); name = argv[0];
#else  
	name = _basename(argv[0]);
#endif

#if 0
	/* this allows the different tools to be called as "mtools -c <command>"
	** where <command> is mdir, mdel, mcopy etcetera
	** Mainly done for the BeOS, which doesn't support links yet.
	*/

	if(argc >= 3 && 
	   !strcmp(argv[1], "-c") &&
	   !strcmp(name, "mtools")) {
		argc-=2;
		argv+=2;
		name = argv[0];
	}
#endif

	/* print the version */
	if(argc >= 2 && 
	   (strcmp(argv[1], "-V") == 0 || strcmp(argv[1], "--version") ==0)) {
		printf("%c%s version %s, dated %s\n", 
		       toupper(name[0]), name+1,
		       mversion, mdate);
		printf("configured with the following options: ");
#ifdef USE_XDF
		printf("enable-xdf ");
#else
		printf("disable-xdf ");
#endif
#ifdef USING_VOLD
		printf("enable-vold ");
#else
		printf("disable-vold ");
#endif
#ifdef USING_NEW_VOLD
		printf("enable-new-vold ");
#else
		printf("disable-new-vold ");
#endif
#ifdef DEBUG
		printf("enable-debug ");
#else
		printf("disable-debug ");
#endif
#ifdef USE_RAWTERM
		printf("enable-raw-term ");
#else
		printf("disable-raw-term ");
#endif
		printf("\n");
		return 0;
	}

	if (argc >= 2 && strcmp(name, "mtools") == 0) {
		/* mtools command ... */
		argc--;
		argv++;
		name = argv[0];
	}
	progname = argv[0];

	read_config();
	setup_signal();
	for (i = 0; i < NDISPATCH; i++) {
		if (!strcmp(name,dispatch[i].cmd)
		    || (name[0] == 'm' && !strcmp(name+1,dispatch[i].cmd)))
			dispatch[i].fn(argc, argv, dispatch[i].type);
	}
	if (strcmp(name,"mtools"))
		fprintf(stderr,"Unknown mtools command '%s'\n",name);
	fprintf(stderr,"Usage: mtools [-V] command [-options] arguments ...\n");
	fprintf(stderr,"Supported commands:");
	for (i = 0; i < NDISPATCH; i++) {
		fprintf(stderr, i%8 == 0 ? "\n\t" : ", ");
		fprintf(stderr, "%s", dispatch[i].cmd);
	}
	putc('\n', stderr);
	fprintf(stderr, "Use 'mtools command -?' for help per command\n");

	return 1;
}
